### ISSUE TITLE
SMPL-X to SMPL beta conversion with transfer_model results in extreme beta values and high loss

### ISSUE THREAD

**Issue by danielrod08** [vchoutas/smplx#226]
# SMPL-X to SMPL Transfer Produces Extreme Beta Values with High Vertex Loss

## Problem Description

I am trying to use the official `transfer_model` to convert SMPL-X beta parameters to SMPL beta parameters.

Despite following the documentation and community suggestions (namedtuple output fix, L2 regularization, beta clamping, adjusting optimization settings, and check the genders), the resulting SMPL betas are still extremely large (e.g., values like `[-17.12, -4.54, -17.84, 0.93, -23.02, 21.12, -6.79, -2.49, -37.11, -30.52]`), and the vertex-to-vertex loss remains high (~500–600 mm) and the final mesh in SMPL isn't human like (I will provide a screenshot with an example).

I have followed all the steps in the README.md and also verified that I am using the neutral gender.

## Code and Configuration

Below I have provided the shell script that I am using to orchestrate the conversion as well as the script that I am using to generate the mesh in smpl.
This experiment was conducted using the default code and configurations provided by the official SMPL-X package, without any custom modifications to the core transfer process

### Conversion Script
**File:** `convert_smplx_to_smpl.sh`

<details>
<summary>Click to expand conversion script</summary>

```bash
#!/bin/bash

# Convert SMPL-X betas to SMPL betas using official transfer_model

# Color codes
GREEN='\033[0;32m'
RED='\033[0;31m'
YELLOW='\033[1;33m'
NC='\033[0m'

print_status() {
    echo -e "${GREEN}[STATUS]${NC} $1"
}

print_error() {
    echo -e "${RED}[ERROR]${NC} $1"
}

# Check arguments
if [ $# -lt 1 ] || [ $# -gt 2 ]; then
    echo "Usage: $0 <input_smplx_betas.npy> [output_smpl_betas.npy]"
    echo "Example: $0 smplx_betas.npy"
    echo "Example: $0 smplx_betas.npy smpl_betas.npy"
    exit 1
fi

INPUT_BETAS="$1"
OUTPUT_BETAS="${2:-smpl_betas.npy}"

# Check if input file exists
if [ ! -f "$INPUT_BETAS" ]; then
    print_error "Input file not found: $INPUT_BETAS"
    exit 1
fi

# Activate conda environment
print_status "Activating smplx_adapter environment..."
eval "$(conda shell.bash hook)"
conda activate smplx_adapter

# Create temporary working directory
WORK_DIR="$(pwd)/temp_processing/transfer_$(date +%s)"
mkdir -p "$WORK_DIR/meshes/smplx"
mkdir -p "$WORK_DIR/output"

print_status "Working directory: $WORK_DIR"

# Step 1: Convert SMPL-X betas to mesh
print_status "Creating SMPL-X mesh from betas..."
python write_obj_single.py "$INPUT_BETAS" "$WORK_DIR/meshes/smplx/shape.obj"

if [ $? -ne 0 ]; then
    print_error "Failed to create mesh from betas"
    rm -rf "$WORK_DIR"
    exit 1
fi

cd smplx

# Step 2: Create config file for transfer_model
print_status "Creating config file..."
cat > "config.yaml" << EOF
datasets:
    mesh_folder:
        data_folder: '$WORK_DIR/meshes/smplx'
output_folder: $WORK_DIR/output
deformation_transfer_path: 'transfer_data/smplx2smpl_deftrafo_setup.pkl'
mask_ids_fname: ''
summary_steps: 100

edge_fitting:
    per_part: False

optim:
    type: 'lbfgs'
    maxiters: 200
    gtol: 1e-06

body_model:
    model_type: "smpl"
    gender: "neutral"
    ext: 'pkl'
    folder: "transfer_data/body_models"
    use_compressed: False
    use_face_contour: True
    smpl:
        betas:
            num: 10

EOF

# Step 3: Run transfer_model
print_status "Running official transfer_model..."
python -m transfer_model --exp-cfg "config.yaml"
TRANSFER_RESULT=$?
cd ..

if [ $TRANSFER_RESULT -ne 0 ]; then
    print_error "Transfer model failed"
    print_error "Check that you have:"
    echo "  - Model files in models/"
    echo "  - Transfer data in transfer_data/"
    echo "  - All dependencies installed"
    rm -rf "$WORK_DIR"
    exit 1
fi

# Step 4: Extract SMPL betas from output
print_status "Extracting SMPL betas..."
python extract_betas.py "$WORK_DIR/output/shape.pkl" "$OUTPUT_BETAS"

if [ $? -ne 0 ]; then
    print_error "Failed to extract betas"
    #rm -rf "$WORK_DIR"
    exit 1
fi

# Clean up
print_status "Cleaning up..."
rm -rf "$WORK_DIR"

print_status "Conversion complete!"
print_status "SMPL betas saved to: $OUTPUT_BETAS"
```
</details>

### Setup Script
**File:** `setup.sh`

<details>
<summary>Click to expand setup script</summary>

```bash
#!/bin/bash

# Setup script for SMPL-X to SMPL adapter using official transfer_model

# Color codes
GREEN='\033[0;32m'
RED='\033[0;31m'
YELLOW='\033[1;33m'
NC='\033[0m'

print_status() {
    echo -e "${GREEN}[SETUP]${NC} $1"
}

print_error() {
    echo -e "${RED}[ERROR]${NC} $1"
}

print_warning() {
    echo -e "${YELLOW}[WARNING]${NC} $1"
}

# Check if conda is available
if ! command -v conda &> /dev/null; then
    print_error "conda not found. Please install Anaconda/Miniconda first."
    exit 1
fi

# Step 1: Create minimal conda environment
#print_status "Creating smplx_adapter environment..."
#conda create -n smplx_adapter python=3.8 -y

# Activate environment
eval "$(conda shell.bash hook)"
conda activate smplx_adapter

# Step 2: Install PyTorch (adjust CUDA version as needed)
#print_status "Installing PyTorch..."
#conda install pytorch torchvision cudatoolkit=11.7 -c pytorch -y

# Step 3: Clone SMPL-X repository
print_status "Cloning SMPL-X repository..."
if [ ! -d "smplx" ]; then
    git clone https://github.com/vchoutas/smplx.git
else
    print_warning "smplx directory already exists, pulling latest changes..."
    cd smplx && git pull && cd ..
fi

# Step 4: Install requirements
print_status "Installing SMPL-X requirements..."
cd smplx
pip install -r requirements.txt
pip install -e .
cd ..

# Step 5: Install additional dependencies
print_status "Installing additional dependencies..."
pip install loguru omegaconf

# Step 6: Install mesh library
print_status "Installing mesh library..."
if [ ! -d "mesh" ]; then
    git clone https://github.com/MPI-IS/mesh.git
    cd mesh
    make all
    cd ..
else
    print_warning "mesh directory already exists"
fi

# Step 7: Install torch-trust-ncg
print_status "Installing torch-trust-ncg optimizer..."
if [ ! -d "torch-trust-ncg" ]; then
    git clone https://github.com/vchoutas/torch-trust-ncg.git
    cd torch-trust-ncg
    pip install -e .
    cd ..
else
    print_warning "torch-trust-ncg directory already exists"
fi

# Step 8: Create directory structure
print_status "Creating directory structure..."
mkdir -p models/smpl
mkdir -p models/smplx
mkdir -p transfer_data
mkdir -p temp_processing

# Step 9: Create helper scripts
print_status "Creating helper scripts..."

# Create write_obj_single.py for converting betas to mesh
cat > write_obj_single.py << 'EOF'
import sys
import numpy as np
import torch
import smplx
import trimesh

def create_mesh_from_betas(betas_path, output_path, model_folder='models', gender='neutral'):
    """Create a mesh from SMPL-X beta parameters."""
    
    # Load betas
    betas = np.load(betas_path)
    print(f"Loaded betas: {betas}")
    
    # Create SMPL-X model
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    model = smplx.create(
        model_folder,
        model_type='smplx',
        gender=gender,
        num_betas=10,
        use_face_contour=False,
        ext='pkl'
    ).to(device)
    
    # Generate mesh with neutral pose
    betas_tensor = torch.tensor(betas, dtype=torch.float32).unsqueeze(0).to(device)
    output = model(
        betas=betas_tensor,
        body_pose=torch.zeros(1, 63, device=device),
        global_orient=torch.zeros(1, 3, device=device),
        transl=torch.zeros(1, 3, device=device),
        return_verts=True
    )
    
    # Save mesh
    vertices = output.vertices.detach().cpu().numpy().squeeze()
    mesh = trimesh.Trimesh(vertices=vertices, faces=model.faces)
    mesh.export(output_path)
    print(f"Saved mesh to {output_path}")

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python write_obj_single.py input_betas.npy output_mesh.obj")
        sys.exit(1)
    
    create_mesh_from_betas(sys.argv[1], sys.argv[2])
EOF

# Create extract_betas.py for extracting betas from pkl
cat > extract_betas.py << 'EOF'
import sys
import pickle
import numpy as np

def extract_betas_from_pkl(pkl_path, output_path):
    """Extract SMPL betas from transfer_model output pkl."""
    
    with open(pkl_path, 'rb') as f:
        data = pickle.load(f, encoding='latin-1')
    
    if 'betas' in data:
        betas = data['betas'].squeeze()
        np.save(output_path, betas)
        print(f"Extracted betas: {betas}")
        print(f"Saved to {output_path}")
    else:
        print("Error: No betas found in pkl file")
        print(f"Available keys: {list(data.keys())}")
        sys.exit(1)

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python extract_betas.py input.pkl output_betas.npy")
        sys.exit(1)
    
    extract_betas_from_pkl(sys.argv[1], sys.argv[2])
EOF

print_status "Setup complete!"
print_status "Next steps:"
echo "1. Download model files from https://smpl-x.is.tue.mpg.de/"
echo "   - Place SMPL models in models/smpl/"
echo "   - Place SMPL-X models in models/smplx/"
echo "2. Download and extract model correspondences to transfer_data/"
echo "3. Use ./convert_smplx_to_smpl.sh to convert SMPL-X betas to SMPL"
```
</details>

### Log

<details>
<summary>Click to expand the log</summary>

```
(smplx_adapter) rod6brg@brg-dl02-w001:~/ATLAS_WP3_2_HMG/shape/adapter$ bash convert_bodyshapegpt.sh "tall" "/home/rod6brg/generated_params/tall.npy"
[STATUS] Generating SMPL-X betas from description...
[STATUS] Description: tall
[Load model start]
Loading checkpoint shards: 100%|████████████████████████████████████████████████████████| 4/4 [00:08<00:00,  2.07s/it]
[Load model complete]
[Inference start]
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[Inference complete]
Description: tall | SMPL-X Shape parameters: [0.91, 0.194, -0.337, 0.749, -1.003, 1.077, 0.201, 0.107, 1.047, -1.097]
Shape parameters saved to: /tmp/tmp.gJoITRSDTB/smplx_betas.npy
[STATUS] Converting SMPL-X betas to SMPL...
[STATUS] Activating smplx_adapter environment...
[STATUS] Working directory: /home/rod6brg/ATLAS_WP3_2_HMG/shape/adapter/temp_processing/transfer_1749382226
[STATUS] Creating SMPL-X mesh from betas...
Loaded betas: [ 0.91   0.194 -0.337  0.749 -1.003  1.077  0.201  0.107  1.047 -1.097]
Gender: neutral
Saved mesh to /home/rod6brg/ATLAS_WP3_2_HMG/shape/adapter/temp_processing/transfer_1749382226/meshes/smplx/shape.obj
[STATUS] Creating config file...
[STATUS] Running official transfer_model...
2025-06-08 12:30:36.849 | INFO     | __main__:main:54 - Saving output to: /home/rod6brg/ATLAS_WP3_2_HMG/shape/adapter/temp_processing/transfer_1749382226/output
WARNING: You are using a SMPL model, with only 10 shape coefficients.
num_betas=10, shapedirs.shape=(6890, 3, 10), self.SHAPE_SPACE_DIM=300
2025-06-08 12:30:36.999 | INFO     | __main__:main:59 - SMPLLayer(
  Gender: NEUTRAL
  Number of joints: 24
  Betas: 10
  (vertex_joint_selector): VertexJointSelector()
)
2025-06-08 12:30:37.848 | INFO     | transfer_model.utils.def_transfer:read_deformation_transfer:40 - Loading deformation transfer from: transfer_data/smplx2smpl_deftrafo_setup.pkl
2025-06-08 12:30:38.220 | WARNING  | __main__:main:74 - Mask ids fname not found:
2025-06-08 12:30:38.221 | INFO     | transfer_model.data.build:build_dataloader:33 - data_folder: /home/rod6brg/ATLAS_WP3_2_HMG/shape/adapter/temp_processing/transfer_1749382226/meshes/smplx

2025-06-08 12:30:38.221 | INFO     | transfer_model.data.datasets.mesh:__init__:47 - Building mesh folder dataset for folder: /home/rod6brg/ATLAS_WP3_2_HMG/shape/adapter/temp_processing/transfer_1749382226/meshes/smplx
2025-06-08 12:30:38.222 | INFO     | transfer_model.data.build:build_dataloader:41 - Creating dataloader with B=1, workers=0
2025-06-08 12:30:38.783 | INFO     | transfer_model.losses.losses:__init__:89 - Building edge loss with norm_type=l2, reduction=mean,
2025-06-08 12:30:38.785 | INFO     | transfer_model.transfer_model:run_fitting:306 - Per-part: False
2025-06-08 12:30:38.785 | INFO     | transfer_model.optimizers.optim_factory:build_optimizer:35 - Building: Lbfgs
2025-06-08 12:30:39.332 | INFO     | transfer_model.optimizers.minimize:minimize:70 - [00000] Loss: 12311.5176
2025-06-08 12:30:39.335 | INFO     | transfer_model.optimizers.minimize:minimize:74 - [00000] Vertex-to-Vertex: 474.0617
2025-06-08 12:30:50.123 | INFO     | transfer_model.optimizers.minimize:minimize:70 - [00100] Loss: 12285.1240
2025-06-08 12:30:50.126 | INFO     | transfer_model.optimizers.minimize:minimize:74 - [00100] Vertex-to-Vertex: 546.9271
Fitting iterations: 100%|███████████████████████████████████████████████████████████| 200/200 [00:18<00:00, 10.73it/s]
2025-06-08 12:30:57.429 | INFO     | transfer_model.optimizers.minimize:minimize:80 - [00200] Loss: 12285.1240
2025-06-08 12:30:57.432 | INFO     | transfer_model.optimizers.minimize:minimize:84 - [00200] Vertex-to-Vertex: 546.9271
2025-06-08 12:30:57.433 | INFO     | transfer_model.optimizers.optim_factory:build_optimizer:35 - Building: Lbfgs
2025-06-08 12:30:57.932 | INFO     | transfer_model.optimizer

<!-- Failed to upload "imagem.png" -->

s.minimize:minimize:70 - [00000] Loss: 2597.2690
2025-06-08 12:30:57.935 | INFO     | transfer_model.optimizers.minimize:minimize:74 - [00000] Vertex-to-Vertex: 468.3900
2025-06-08 12:31:24.569 | INFO     | transfer_model.optimizers.minimize:minimize:70 - [00100] Loss: 2070.1865
2025-06-08 12:31:24.573 | INFO     | transfer_model.optimizers.minimize:minimize:74 - [00100] Vertex-to-Vertex: 460.4836
Fitting iterations: 100%|███████████████████████████████████████████████████████████| 200/200 [00:39<00:00,  5.07it/s]
2025-06-08 12:31:36.872 | INFO     | transfer_model.optimizers.minimize:minimize:80 - [00200] Loss: 2070.1865
2025-06-08 12:31:36.875 | INFO     | transfer_model.optimizers.minimize:minimize:84 - [00200] Vertex-to-Vertex: 460.4836
100%|███████████████████████████████████████████████████████████████████████████████████| 1/1 [00:58<00:00, 58.81s/it]
[STATUS] Extracting SMPL betas...
Extracted betas: [-17.134598   -3.3211298  -6.7204814   2.9502811 -29.490187   -9.590364
  -5.0027404  13.738536    8.658828   38.82055  ]
Saved to /tmp/tmp.gJoITRSDTB/smpl_betas.npy
[STATUS] Cleaning up...
[STATUS] Conversion complete!
[STATUS] SMPL betas saved to: /tmp/tmp.gJoITRSDTB/smpl_betas.npy
[STATUS] Pipeline complete!
[STATUS] SMPL betas saved to: /home/rod6brg/generated_params/tall.npy
SMPL betas: [-17.134598   -3.3211298  -6.7204814   2.9502811 -29.490187   -9.590364
  -5.0027404  13.738536    8.658828   38.82055  ]
```
</details>

### Create a SMPL-X mesh
**File:** `write_obj_single.py`

<details>
<summary>Click to expand script</summary>

```python
import sys
import numpy as np
import torch
import smplx
import trimesh

def create_mesh_from_betas(betas_path, output_path, model_folder='models', gender='neutral'):
    """Create a mesh from SMPL-X beta parameters."""
    
    # Load betas
    betas = np.load(betas_path)
    print(f"Loaded betas: {betas}")
    print(f"Gender: {gender}")
    # Create SMPL-X model
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    model = smplx.create(
        model_path=model_folder,
        model_type='smplx',
        gender=gender,
        num_betas=10,
        num_expression_coeffs=10,
        use_face_contour=False,
        ext='pkl'
    ).to(device)
    
    # Generate mesh with neutral pose
    betas_tensor = torch.tensor(betas, dtype=torch.float32).unsqueeze(0).to(device)
    
    output = model(
        betas=betas_tensor,
        body_pose=torch.zeros(1, 63, device=device),
        expression=torch.zeros(1, 10, device=device),
        global_orient=torch.zeros(1, 3, device=device),
        transl=torch.zeros(1, 3, device=device),
        return_verts=True
    )
    
    # Save mesh
    vertices = output.vertices.detach().cpu().numpy().squeeze()
    mesh = trimesh.Trimesh(vertices=vertices, faces=model.faces)
    mesh.export(output_path)
    print(f"Saved mesh to {output_path}")

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python write_obj_single.py input_betas.npy output_mesh.obj")
        sys.exit(1)
    
    create_mesh_from_betas(sys.argv[1], sys.argv[2])

```
</details>

![Image](https://github.com/user-attachments/assets/fb287bf5-4d51-47a7-a1ed-c00562cc515c)

 Thank you very much for your attention and help!
